orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -785,10 +907,19 @@ Status TraverseGraphWithConnectedElement(Graph& graph,
                                          std::set<NodeArg*>& visited_inputs,
                                          std::set<NodeArg*>& visited_outputs) {
   assert(start_node);
+
   visited_nodes.clear();
   visited_inputs.clear();
   visited_outputs.clear();
 
+  for (const auto node_arg : graph.GetInputs()) {
+    if (!node_arg->Exists()) {
+      continue;
+    }
+    NodeArg* mutable_node_arg = graph.GetNodeArg(node_arg->Name());
+    visited_inputs.insert(mutable_node_arg);
+  }
```

Comment:
This logic doesn't seem right to me -- not all partitions connects to all inputs. In fact, in most cases, only the first partition is connected to input, and the following partitions just get tensors from `recv` node. If the purpose of this logic is for shape inferencing and not require any graph traversing at all, we can move this piece outside the traversing function and add a comment about this operation.

Reply:
In addition to Recv, intermediate and last partitions have events, loss scale, learning rate, etc., so we make all inputs available to all stages.

---
In reply to: [538776120](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r538776120) [](ancestors = 538776120)